### ISSUE TITLE
Plugins support

### ISSUE THREAD

**Issue by gedw99** [Drakirus/go-flutter-desktop-embedder#10]
These are two desktop services I am thinking of adding and I want feedback

1. Notifications.
There are good golang packages for this so should be easy to do native notification.

2. Unilinks is when a user gets an email with a deeplinking into the flutter app. When they click the link it opens the flutter app and routes into the link.
Unilinks is the name of the flutter for mobile plugin and so we can reuse the Dart code and replace the java and objective c with golang. I am not sure what's already out there in terms of golang libs for this but I am confident it's doable.

3. Printing
Dart has pdf gen but golang has great pdf gen.  Assuming we want to go with the PDF approach ? Other way is to take a screen Scape raster image and bundle it inside a pdf.  Both are useful.
Then there is sending it to a printer. There are some decent golang libs for this.

If I can get feedback, and I will start working on these.

**Comment by Drakirus** [comment]
Again, Thanks for your suggestion :100:,

1. Notifications should be easy. (Plugin required).
2. Unilinks is harder but doable (Plugin required).
 The [Fluro routing library](https://github.com/theyakka/fluro/blob/d19e68935692b9410f26afd861325394bcd0128b/example/lib/config/route_handlers.dart#L69) is doing something similar on the `Android -> Flutter` side. If Unilinks gets ported having the _GoFlutter embedder_ compatible with _Fluro_ would be the best. (I have no idea how the `Android intent` is handled on the embedder side)
3. Printing 
    -  **PDF Gen** -  if the generation of a PDF can be done on the Flutter side it should say on the Flutter side. (The _flutter/engine_ can save files to the drive without having to go through a plugin)
    -  _"take a screen Scape raster image and bundle it inside a pdf."_ I don’t understand, can you clarify?
    -  Printing to a printer should be easy to handle (And not only PDF, images,..) (Plugin required).

Before implementing those desktop services, we need 
1. to have a 'Golang way' to handle the _flutter/engine_ plugin message request and response.
2. a way for the user to register they plugin into the _GoFlutter embedder_.
    -   It should be modular
    -   Do the user have to recompile the binaries in order for their plugin to work or can the plugin be registered at runtime?

**Comment by gedw99** [comment]
Did you see this ?

https://github.com/adieu/flutter_go

It's a way to plugin golang into flutter on mobile.

It's simple and works with zero messing around with java or objective c.

This is what might be cool for desktop I was thinking.

Then using g your your golang lib between mobile and desktop is the same

Food for thought ?

**Comment by gedw99** [comment]
now there is a grpc approach also.
https://github.com/empirefox

Quite nice since it supports both gomobile and go. So can be used for Mobiles and Desktops.

**Comment by Drakirus** [comment]
Those golang libs are looking good. But they are kind of niche. I don't see many cases where you would use those 'codec' instead of the ones that are already exposed.

A lot of work needs to be done to implement the `StandardMethodCodec` and `JSONMethodCodec` which are used by most of the library maker. The goal is to support those codecs so that we can plug the embedder to already existing flutter libs.

**Comment by gedw99** [comment]
Did you see the JSON RPC one ?

Here is my make file for it:

````

ROOTLIB=github.com/renatoathaydes
ROOTLIB_FSPATH=$(GOPATH)/src/$(ROOTLIB)

LIB=github.com/renatoathaydes/go-hash
LIB_FSPATH=$(GOPATH)/src/$(LIB)

MOB_LIB=github.com/renatoathaydes/gohash_mobile
MOB_LIB_FSPATH=$(GOPATH)/src/$(MOB_LIB)

FL_LIB=github.com/renatoathaydes/gohash_mobile_app
FL_LIB_FSPATH=$(GOPATH)/src/$(FL_LIB)


print:
	@echo

	@echo ROOTLIB  -
	@echo ROOTLIB :$(ROOTLIB)
	@echo ROOTLIB_FSPATH  : $(ROOTLIB_FSPATH)
	#cd $(LIB_FSPATH) && ls -al 
	@echo

	@echo LIB  -
	@echo LIB :$(LIB)
	@echo LIB_FSPATH  : $(LIB_FSPATH)
	#cd $(LIB_FSPATH) && ls -al 
	@echo

	@echo MOB_LIB  -
	@echo MOB_LIB :$(MOB_LIB)
	@echo MOB_LIB_FSPATH  : $(MOB_LIB_FSPATH)
	#cd $(LIB_FSPATH) && ls -al 
	@echo

	@echo Flutter LIB  -
	@echo FL_LIB :$(FL_LIB)
	@echo FL_LIB_FSPATH  : $(FL_LIB_FSPATH)
	#cd $(LIB_FSPATH) && ls -al 
	@echo


	@echo

dep:
	cd $(ROOTLIB_FSPATH) && git clone https://$(LIB).git
	cd $(ROOTLIB_FSPATH) && git clone https://$(MOB_LIB).git
	cd $(ROOTLIB_FSPATH) && git clone https://$(FL_LIB).git

dep-os:
	# Setup gomobile properly
	# gomobile init. Must use the NDK bundle 
	gomobile init -ndk $(ANDROID_NDK)

dep-update:
	# get latest code from git
	cd $(LIB_FSPATH) && git pull
	cd $(MOB_LIB_FSPATH) && git pull
	cd $(LIB_FSPATH) && git pull

	# update its deps ( needed due to golang 1.11 )
	#cd $(LIB_FSPATH) && dep ensure -update

	# update flutter code
	cd $(MOB_LIB_FSPATH) && flutter packages upgrade
	cd $(FL_LIB_FSPATH) && flutter packages upgrade

dep-clean:
	cd $(ROOTLIB_FSPATH) && rm -rf *

code:
	code $(ROOTLIB_FSPATH)

###

ex-run:
	flutter emulators --launch Nexus_5X_API_27
	flutter emulators --launch apple_ios_simulator

	cd $(MOB_LIB_FSPATH)/example && flutter run -d all

ex-build:
	# build desktop libs
	#cd $(LIB_FSPATH) && make release

	# build mobile libs
	# this is not needed for Flutter because gradle is setup to do it.
	#gomobile bind -target=android $(LIB)/mobileapi
	#gomobile bind -target=ios $(LIB)/mobileapi
	
	cd $(MOB_LIB_FSPATH)/example && flutter build apk

	# IOS needs a proper provisioning profile to match the name. Have to make it in the Apple Web gui.
	cd $(MOB_LIB_FSPATH)/example && flutter build ios
	

run:
	flutter emulators --launch Nexus_5X_API_27
	flutter emulators --launch apple_ios_simulator

	cd $(FL_LIB_FSPATH) && flutter run -d all



````

**Comment by Drakirus** [comment]
Yup I did saw it, but for now, I only want to keep the standard related work in this repo. Another golang lib could be made later to bridge their stuff into this project.

**Comment by gedw99** [comment]
ok, makes sense :)

**Comment by Drakirus** [comment]
I just made a PR #20 
Any through ?

**Comment by Drakirus** [comment]
Fixed with: #20 and #25
